### PR TITLE
[elixir-omg] Add support for zero touch Watcher deployments

### PR DESCRIPTION
**Pull request by jbunce**
Adds support for Watcher zero touch deployments.

Pushing before `develop` gets deprecated in favour of OIP4.

### CODE REVIEW

**Comment by jbunce** [comment]
```
2018-12-05 17:13:34,663 INFO    :Starting Launcher
2018-12-05 17:13:34,669 INFO    :Service type to launch is Elixir Childchain
2018-12-05 17:13:34,669 INFO    :Starting launch process for build 0df129893192bb93728188a5297
42798ec9efb0f
2018-12-05 17:13:34,690 INFO    :Connected to the Ethereum client
2018-12-05 17:13:34,690 INFO    :Ethereum client is b'{"jsonrpc":"2.0","id":67,"result":"Geth/
v1.8.20-unstable-f850123e/linux-amd64/go1.11.2"}\n'
omg_watcher: generated priv/static/swagger.json
2018-12-05 17:13:36,943 INFO    :Elixir mix compile successful
2018-12-05 17:13:46,103 INFO    :Smart contract deployed and authority address populated
2018-12-05 17:13:46,110 INFO    :Updated Contract Exchanger with ~./config.exs data
2018-12-05 17:13:48,474 INFO    :Childchain database initialised

elixir-user@childchain-7dcc9c4659-p7z8h:~/elixir-omg$ cat ~/config.exs
use Mix.Config
config :omg_eth,
  contract_addr: "0xf2339ec0dd159bc7e7d485f27fcfafdea93ac4b6",
  txhash_contract: "0x60c04d6178e2105c656c9ed7a5c6138dd056183d4bcc7ee3891898b15701ba39",
  authority_addr: "0xfbbfc66360f2842a34d4faef0e3c3494224ca898"

2018-12-05 17:14:33,276 INFO    :Starting Launcher
2018-12-05 17:14:33,283 INFO    :Service type to launch is Elixir Watcher
2018-12-05 17:14:33,283 INFO    :Starting launch process for build 0df129893192bb93728188a529742798ec9efb0f
2018-12-05 17:14:33,302 INFO    :Connected to the Ethereum client
2018-12-05 17:14:33,302 INFO    :Ethereum client is b'{"jsonrpc":"2.0","id":67,"result":"Geth/v1.8.20-unstab
le-f850123e/linux-amd64/go1.11.2"}\n'
omg_watcher: generated priv/static/swagger.json
2018-12-05 17:14:35,552 INFO    :Elixir mix compile successful
2018-12-05 17:14:35,559 INFO    :Response received from the contract exchanger service
2018-12-05 17:14:35,560 INFO    :Written config_watcher.exs
2018-12-05 17:14:40,793 INFO    :Watcher Postgres instance initialised

elixir-user@watcher-7cfd5c454c-hdcfk:~/elixir-omg$ cat ~/config_watcher.exs
use Mix.Config
config :omg_eth,
  contract_addr: "0xf2339ec0dd159bc7e7d485f27fcfafdea93ac4b6",
  txhash_contract: "0x60c04d6178e2105c656c9ed7a5c6138dd056183d4bcc7ee3891898b15701ba39",
  authority_addr: "0xfbbfc66360f2842a34d4faef0e3c3494224ca898"
config :omg_db,
  leveldb_path: Path.join([System.get_env("HOME"), ".omg/data_watcher"])
```

**Comment by jbunce** [comment]
Supports childchain recovery from a service failure:

```
2018-12-06 08:54:14,559 INFO    :Starting Launcher
2018-12-06 08:54:14,565 INFO    :Service type to launch is Elixir Childchain
2018-12-06 08:54:14,566 INFO    :Starting launch process for build eb204306316b2f6
5bc42c088febf13c59985e92d
2018-12-06 08:54:14,566 INFO    :Chain data not found
2018-12-06 08:54:14,584 INFO    :Connected to the Ethereum client
2018-12-06 08:54:14,584 INFO    :Ethereum client is b'{"jsonrpc":"2.0","id":67,"re
sult":"Geth/v1.8.20-unstable-f850123e/linux-amd64/go1.11.2"}\n'
omg_watcher: generated priv/static/swagger.json
2018-12-06 08:54:17,016 INFO    :Elixir mix compile successful
2018-12-06 08:54:26,109 INFO    :Smart contract deployed and authority address pop
ulated
2018-12-06 08:54:26,118 INFO    :Updated Contract Exchanger with ~./config.exs dat
a
2018-12-06 08:54:28,576 INFO    :Childchain database initialised
2018-12-06 08:54:28,577 INFO    :Launcher process complete
2018-12-06 08:54:31.351 [info] module=OMG.API.Application function=start/2 Started
 application OMG.API.Application
2018-12-06 08:54:31.369 [info] module=OMG.DB function=utxos/1 Reading UTXO set, th
is might take a while. Allowing 600000 ms
2018-12-06 08:54:31.379 [info] module=OMG.API.State function=init/1 Started State,
 height: 0, deposit height: 0
2018-12-06 08:54:31.553 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arting BlockQueue at parent_height: 260704, parent_start: 260699, mined_child_bloc
k: 0, stored_child_top_block: 0
2018-12-06 08:54:31.569 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arting BlockQueue, top_mined_hash: "0000000000000000000000000000000000000000000000
000000000000000000"
2018-12-06 08:54:31.569 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arted BlockQueue
2018-12-06 08:54:31.585 [info] module=OMG.API.FeeChecker function=update_fee_spec/
0 Reloaded 1 fee specs from file, changed at 1540626006
2018-12-06 08:54:31.585 [info] module=OMG.API.FeeChecker function=init/1 Started F
eeChecker
2018-12-06 08:54:31.598 [info] module=OMG.API.EthereumEventListener function=init/
1 Starting EthereumEventListener for depositor
2018-12-06 08:54:31.608 [info] module=OMG.API.EthereumEventListener function=init/
1 Starting EthereumEventListener for exiter


2018-12-06 08:55:01,350 INFO    :Starting Launcher
2018-12-06 08:55:01,356 INFO    :Service type to launch is Elixir Childchain
2018-12-06 08:55:01,356 INFO    :Starting launch process for build eb204306316b2f6
5bc42c088febf13c59985e92d
2018-12-06 08:55:01,356 INFO    :Childchain data found
2018-12-06 08:55:01,376 INFO    :Connected to the Ethereum client
2018-12-06 08:55:01,376 INFO    :Ethereum client is b'{"jsonrpc":"2.0","id":67,"re
sult":"Geth/v1.8.20-unstable-f850123e/linux-amd64/go1.11.2"}\n'
2018-12-06 08:55:01,383 INFO    :Response received from the contract exchanger ser
vice
2018-12-06 08:55:01,383 INFO    :Written config.exs
2018-12-06 08:55:01,383 INFO    :Launcher process complete
omg_watcher: generated priv/static/swagger.json
2018-12-06 08:55:04.168 [info] module=OMG.API.Application function=start/2 Started
 application OMG.API.Application
2018-12-06 08:55:04.186 [info] module=OMG.DB function=utxos/1 Reading UTXO set, th
is might take a while. Allowing 600000 ms
2018-12-06 08:55:04.197 [info] module=OMG.API.State function=init/1 Started State,
 height: 0, deposit height: 0
2018-12-06 08:55:04.366 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arting BlockQueue at parent_height: 260737, parent_start: 260699, mined_child_bloc
k: 0, stored_child_top_block: 0
2018-12-06 08:55:04.382 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arting BlockQueue, top_mined_hash: "0000000000000000000000000000000000000000000000
000000000000000000"
2018-12-06 08:55:04.382 [info] module=OMG.API.BlockQueue.Server function=init/1 St
arted BlockQueue
2018-12-06 08:55:04.396 [info] module=OMG.API.FeeChecker function=update_fee_spec/
0 Reloaded 1 fee specs from file, changed at 1540626006
2018-12-06 08:55:04.396 [info] module=OMG.API.FeeChecker function=init/1 Started F
eeChecker
2018-12-06 08:55:04.409 [info] module=OMG.API.EthereumEventListener function=init/
1 Starting EthereumEventListener for depositor
2018-12-06 08:55:04.418 [info] module=OMG.API.EthereumEventListener function=init/
1 Starting EthereumEventListener for exiter
```

File: launcher.py [hunk 1]
```diff
@@ -48,6 +117,9 @@ def compile_application(self) -> bool:
     def deploy_contract(self) -> bool:
         ''' Deploy the smart contract and populate the ~/config.exs file
         '''
+        if self.ethereum_network == [x for x in self.public_networks]:
```
Comment (paulperegud) [inline launcher.py]: Shouldn't it be `if self.ethereum_network in self.public_networks:`?

File: launcher.py [hunk 2]
```diff
@@ -65,10 +137,63 @@ def deploy_contract(self) -> bool:
         )
         return False
 
+    def clean_config_entry(self, line: str) -> str:
+        ''' Clean the line that forms the entry for the config.exs
+        '''
+        line = re.sub(r'"', '', line)
```
Comment (paulperegud) [inline launcher.py]: `line.replace('"'. '')` is more conservative approach here

File: launcher.py [hunk 3]
```diff
@@ -65,10 +137,63 @@ def deploy_contract(self) -> bool:
...
+                    contract_details['contract_addr'] = \
+                        self.clean_config_entry(entry.split(' ')[3])
+                if 'txhash_contract' in entry:
+                    contract_details['txhash_contract'] = \
+                        self.clean_config_entry(entry.split(' ')[3])
+                if 'authority_addr' in entry:
+                    contract_details['authority_addr'] = \
+                        self.clean_config_entry(entry.split(' ')[3])
+
+        requests.post(
```
Comment (paulperegud) [inline launcher.py]: This manual parsing of config files is brutal. What do you think about using .toml as a format for `elixir-omg` config files?

File: launcher.py (hunk 2 above)
Comment (jbunce) [reply launcher.py]: Addressed in 41f2545

File: launcher.py (hunk 1 above)
Comment (jbunce) [reply launcher.py]: Thanks, that was a bug. Resolved in 41f2545

File: launcher.py (hunk 3 above)
Comment (jbunce) [reply launcher.py]: I agree. Any structured data format would be better than parsing a `config.exs`. Something for after public testnet?

**Pull request merged**